### PR TITLE
add n-dimensional support to denoise_wavelet

### PR DESCRIPTION
**Pull request by grlee77**
## Description
This adds n-dimensional support to `denoise_wavelet`.  The multichannel handling matches the existing `denoise_tv_chambolle` which also has n-dimensional support.

### CODE REVIEW

**Comment by JDWarner** [comment]
:+1: for this approach, and LGTM. This would be how to handle, e.g., several MRI sequences which are inherently independent. 

For color, though, we clearly need something a little different. Working on each channel in isolation actually can exacerbate color noise. That's an issue for a different PR, though (discuss in #2240).

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -394,7 +394,8 @@ def _wavelet_threshold(img, wavelet, threshold=None, sigma=None, mode='soft'):
     return pywt.waverecn(denoised_coeffs, wavelet)
 
 
-def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft',
+                    multichannel=False):
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: In the docstring, please, change expected data type for `img` to smth like `(M[, N[, ..., P]][, C])`, and add description for `multichannel` parameter.

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -457,16 +458,14 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
 
     img = img_as_float(img)
 
-    if img.ndim not in {2, 3}:
-        raise ValueError('denoise_wavelet only supports 2D and 3D images')
-
-    if img.ndim == 2:
+    if multichannel:
+        out = np.empty_like(img)
+        for c in range(img.ndim):
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: This is not correct for 4+D case, use `img.shape[-1]` instead.

Comment (grlee77) [inline skimage/restoration/_denoise.py]: good catch.  clearly I had not tested the n-d case yet!

**Comment by grlee77** [comment]
I plan to update the tests to add a 4D test case.  
Also, I still have to make sure any examples using this have the `multichannel` argument set appropriately.

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -400,7 +400,7 @@ def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft',
 
     Parameters
     ----------
-    img : ndarray (2D/3D) of ints, uints or floats
+    img : ndarray ([M[, N[, ...P]][, C]) of ints, uints or floats
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: Sorry, actually this should be `(M, N[, ...P][, C])` if we declare 2+D support. Please, notice that `[, ]` mark _optional_ dimensions.

**Comment by soupault** [comment]
One last remark, otherwise LGTM.

File: skimage/restoration/tests/test_denoise.py [hunk 4]
```diff
@@ -310,16 +310,20 @@ def test_no_denoising_for_small_h():
...
         assert denoised.sum()**2 <= img.sum()**2
 
         # test changing noise_std (higher threshold, so less energy in signal)
-        assert (restoration.denoise_wavelet(noisy, sigma=0.2).sum()**2 <=
-                restoration.denoise_wavelet(noisy, sigma=0.1).sum()**2)
+        res1 = restoration.denoise_wavelet(noisy, sigma=0.2,
+                                           multichannel=multichannel)
+        res2 = restoration.denoise_wavelet(noisy, sigma=0.1,
+                                           multichannel=multichannel)
+        assert (res1.sum()**2 <= res2.sum()**2)
```
Comment (jni) [inline skimage/restoration/tests/test_denoise.py]: I know this wasn't your approach, but this test seems quite weak to me! How about showing a reduction in the noise after denoise_wavelet, or increase in SNR?

**Comment by jni** [comment]
@grlee77 thank you! This PR makes me very happy. What do you think of my test suggestion? We do have a policy that new PRs shouldn't have to fix old code issues, so feel free to ignore.

File: skimage/restoration/tests/test_denoise.py (hunk 4 above)
Comment (grlee77) [inline skimage/restoration/tests/test_denoise.py]: sure.  it should be simple to verify at least a rough improvement via `compare_psnr`

File: skimage/restoration/_denoise.py (hunk 3 above)
Comment (grlee77) [inline skimage/restoration/_denoise.py]: I think 1D is already supported here, though?

**Comment by grlee77** [comment]
Tests for 1D through 4D inputs were added and PSNR is compared.  
The `plot_denoise.py` example has also been updated.

PSNR only improves slightly with the default sigma (~10%).  To get a reasonably denoised image, the user has to specify sigma larger than the true value. This was resolved in #2241, so I will not plan to address that in this PR.

File: skimage/restoration/_denoise.py (hunk 3 above)
Comment (soupault) [inline skimage/restoration/_denoise.py]: True. Let it be `(M[, N[, ...P]][, C])` then.

**Comment by jni** [comment]
Brilliant! Thank you, @grlee77!

**Pull request merged**